Add steps to use ESC with Pulumi IaC (#3482)
pulumi/registry: themes/default/content/registry/packages/aws/installation-configuration.md

| `skipRegionValidation` | Optional | Skip static validation of region name. Used by users of alternative AWS-like APIs or users w/ access to regions that are not public (yet). |
| `skipRequestingAccountId` | Optional | Skip requesting the account ID. Used for AWS API implementations that do not have IAM/STS API and/or metadata API. |
| `token` | Optional | Use this to set an MFA token. It can also be sourced from the `AWS_SESSION_TOKEN` environment variable. |

> ## Centralize your configuration

Reviewer: Since this is an option for configuring AWS, what do you think about listing it under the Configuration section of this page? I ask because the paragraph that introduces that section says, "There are a few different ways you can configure your AWS credentials to work with Pulumi", and this is another one of those ways. What do you think? Any reason not to?

Author: I was initially thinking the same thing as far as putting it in that area, but I felt a bit better about it being in its own section given the flow of the content. There's the section right above it that talks about the different configuration options and I like that we can say "hey it's not just the AWS environment variables, you can centralize these other config options we mentioned too". Also I like that users can navigate the different sub sections in this section using the left hand menu which they wouldn't see if it was included further above. I wouldn't say this is a hill I'm willing to die on though lol happy to chat about it

Reviewer: Yeah, we should talk through it. In addition to just "keeping all the configuration options together", this approach could also help in preventing ESC from appearing too complicated. Just having more content can make it look like "whoa, I don't want to take all that on", when in fact (especially with AWS) it's actually pretty simple to set up and use. So to the extent we can make this stuff appear as compact and simple as we can, it'll probably help encourage users to give it a try.

> While you can create your AWS configurations locally, you also have the option to centralize your configurations using [Pulumi ESC (Environments, Secrets, and Configuration)](https://www.pulumi.com/docs/pulumi-cloud/esc/). Using this service will enable you to do things like run AWS CLI commands or run the `pulumi up` command with dynamically generated credentials, removing the need to configure and manage your credentials locally.

Reviewer: ```suggestion
In addition to configuring the AWS provider locally, you also have the option to centralize your configurations using [Pulumi ESC (Environments, Secrets, and Configuration)](/docs/pulumi-cloud/esc/). Using this service will enable you to run AWS or Pulumi CLI commands with dynamically generated credentials, removing the need to configure and manage your credentials locally.
```

To do this, you will need to complete the following steps:

### Configure OIDC between Pulumi and AWS

Refer to the [Configuring OpenID Connect for AWS Guide](https://www.pulumi.com/docs/pulumi-cloud/oidc/aws/) for the step-by-step process on how to do this. Note that when adding your configuration to your environment file, you can also define your AWS region in the `environmentVariables` section:

```yaml
values:
  aws:
    login:
      fn::open::aws-login:
        oidc:
          duration: 1h
          roleArn: <your-oidc-iam-role-arn>
          sessionName: pulumi-environments-session
  environmentVariables:
    AWS_ACCESS_KEY_ID: ${aws.login.accessKeyId}
    AWS_SECRET_ACCESS_KEY: ${aws.login.secretAccessKey}
    AWS_SESSION_TOKEN: ${aws.login.sessionToken}
    AWS_REGION: <YOUR_AWS_REGION> # e.g.`ap-south-1`
```

> ### [Optional] Move Pulumi config to your ESC environment

Reviewer: Is a separate section needed here? I ask because this section is really only conveying that you can add a `pulumiConfig` block if you want -- which, I now that I think of it, could also be used to apply _all_ of these settings (ID, key, token, and region), rather than just region. Right? This, for example, is another way of saying the same thing:

```yaml
# myorg/myenv
aws:
    login:
      fn::open::aws-login:
        oidc:
          duration: 1h
          roleArn: <your-oidc-iam-role-arn>
          sessionName: pulumi-environments-session

values:
  pulumiConfig:
    aws:accessKey: ${aws.login.accessKeyId}
    aws:secretKey: ${aws.login.secretAccessKey}
    aws:token: ${aws.login.sessionToken}
    aws:region: ${aws.region}
```

What do you think about showing both of these in the same section, along with the content of `Import your environment`, since it all goes together (and again is collectively just one more way to configure the AWS provider)?

Author: I'm glad you pointed this out because I was actually thinking of updating this section so that the example I show in the `pulumiConfig` part is not specific to the credentials variables. I want to let folks know that "hey you can include other Pulumi config in here, not just AWS credentials and not just things related to AWS".

I would be happy with something like:

```yaml
values:
  aws:
    login:
      fn::open::aws-login:
        oidc:
          duration: 1h
          roleArn: <your-oidc-iam-role-arn>
          sessionName: pulumi-environments-session
  environmentVariables:
    AWS_ACCESS_KEY_ID: ${aws.login.accessKeyId}
    AWS_SECRET_ACCESS_KEY: ${aws.login.secretAccessKey}
    AWS_SESSION_TOKEN: ${aws.login.sessionToken}
    AWS_REGION: <YOUR_AWS_REGION>
  pulumiConfig: # exposes Pulumi config values to the Pulumi CLI
    project:environment: 'dev'
    aws:roleArn: 'arn:aws:iam::058111598222:role/OrganizationAccountAccessRole'
    aws:dynamodbEndpoint: 'dynamodb.us-east-2.amazonaws.com'
```

[note that adds your comment about not having to expose environment variables if it is not needed for the workflow] For example:

```yaml
values:
  aws:
    login:
      fn::open::aws-login:
        oidc:
          duration: 1h
          roleArn: <your-oidc-iam-role-arn>
          sessionName: pulumi-environments-session
  pulumiConfig: 
    aws:region: <YOUR_AWS_REGION>
    aws:accessKey: ${aws.login.accessKeyId}
    aws:secretKey: ${aws.login.secretAccessKey}
    aws:token: ${aws.login.sessionToken}
    project:environment: 'dev'
    aws:roleArn: 'arn:aws:iam::058111598222:role/OrganizationAccountAccessRole'
    aws:dynamodbEndpoint: 'dynamodb.us-east-2.amazonaws.com'
```

It was mentioned earlier in this guide that you can also set your AWS region as a Pulumi configuration value in your project's stack settings file (`Pulumi.<stack-name>.yaml`). In addition to this, there may be other values that you have defined in this file, and these values can also be centralized using Pulumi ESC. To [expose these values to Pulumi IaC](https://www.pulumi.com/docs/pulumi-cloud/esc/environments/#projecting-pulumi-config), you will need to add a second-level key called `pulumiConfig` and nest any desired values underneath it as shown below:

```yaml
values:
  aws:
    login:
      fn::open::aws-login:
        oidc:
          duration: 1h
          roleArn: <your-oidc-iam-role-arn>
          sessionName: pulumi-environments-session
  environmentVariables:
    AWS_ACCESS_KEY_ID: ${aws.login.accessKeyId}
    AWS_SECRET_ACCESS_KEY: ${aws.login.secretAccessKey}
    AWS_SESSION_TOKEN: ${aws.login.sessionToken}
  pulumiConfig: # exposes Pulumi config values to the Pulumi CLI
    aws:region: <YOUR_AWS_REGION> # e.g.`ap-south-1`
```

> Note that the configuration values under `pulumiConfig` can also be referenced directly from within your Pulumi program code. This is done using the same method to reference values from your project's stack settings file. You can see examples of how to do this in the [Accessing Configuration from Code](https://www.pulumi.com/docs/concepts/config/#code) section of the Pulumi documentation.

Reviewer: Keeping links root-relative wherever possible is best -- we aren't yet linting for this, but we probably should: 

```suggestion
Note that the configuration values under `pulumiConfig` can also be referenced directly from within your Pulumi program code. This is done using the same method to reference values from your project's stack settings file. You can see examples of how to do this in the [Accessing Configuration from Code](/docs/concepts/config/#code) section of the Pulumi documentation.
```

Author: I remember having some issues with the links loading in the shorter format which is why I've done it this way. For some reason I would get a 404 when testing the root relative links. I had assumed it was because the Registry and other docs are now separated? But maybe it was a fluke. I will try again.

Author: ahhhh fair enough, totally missed that. I did add the shortened links and will just work under the assumption that they will work as expected when published.

### Import your environment

The last step is to update your project's stack settings file (`Pulumi.<stack-name>.yaml`) to import your ESC environment as shown below:

```yaml
environment:
  - <your-environment-name>
```

Make sure to replace `<your-environment-name>` with the name of the ESC environment you created in the previous steps.

> You can test that your configuration is working by running the `pulumi up` command. This will deploy your AWS resources using the dynamically generated credentials in your environment file.

Reviewer: Might be better to suggest `preview`, since this won't deploy any resources, and will also fail if any of these values are missing:

```suggestion
You can test that your configuration is working by running `pulumi preview`. 
```

{{< notes type="info" >}}
Make sure that your local environment does not have AWS credentials configured before running this command. You can check by running something like the `aws s3 ls` command which should return the following:

'Unable to locate credentials. You can configure credentials by running "aws configure".'
{{< /notes >}}

> To learn more about projecting environment variables in Pulumi ESC, refer to the [relevant documentation](https://www.pulumi.com/docs/pulumi-cloud/esc/environments/#projecting-environment-variables).

Reviewer 2: ```suggestion
To learn more about projecting environment variables in Pulumi ESC, refer to the [relevant ESC documentation](https://www.pulumi.com/docs/pulumi-cloud/esc/environments/#projecting-environment-variables).
```
